File: python/prisoners.py
#	prisoners

from matrix import *

#Declare global variables
precision = None
yearsInPrison = None

'''
	Function: setPrecision
	Description:
		set the precision value for the game
	Input:
		Precision (Number): new value
'''
def setPrecision(Precision):
	global precision
	precision = Precision
	
'''
	Function: getPrecision
	Description:
		get the precision value for the game
	Output:
		return (Number): current value
'''
def getPrecision(Precision):
	global precision
	return precision

'''
	Function: setYearsInPrison
	Description:
		set the yearsInPrison value for the game
	Input:
		YearsInPrison (List): new matrix
'''
def setYearsInPrison(YearsInPrison):
	global yearsInPrison
	yearsInPrison = YearsInPrison

'''
	Function: getYearsInPrison
	Description:
		get the yearsInPrison value for the game
	Output:
		return (List): current matrix
'''
def getYearsInPrison(YearsInPrison):
	global yearsInPrison
	return yearsInPrison

'''
	Class: prisoner
	Description:
		representation of a player in a game; inherits from player
'''
class prisoner(player):
	def setEmpathy(self,Empathy):
		global precision
		self.Empathy = Empathy
		self.Apathy = precision - Empathy
	def getEmpathy(self):
		return self.Empathy
	def setApathy(self,Apathy):
		global precision
		self.Apathy = Apathy
		self.Empathy = precision - Apathy
	def getApathy(self):
		return self.Apathy

'''
	Function: preferRationalPD
	Description:
		Maps referRational result 0 to 'confess', and result 1 to 'refuse'
	Input:
		Player (prisoner): the player
	Output:
		return (String): 'confess' or 'refuse'
'''
def preferRationalPD(Player):
	deriveUtility(Player)
	if preferRational(Player):
		Action = 'refuse'
	else:
		Action = 'confess'
	return Action

'''
	Function: deriveUtility
	Description:
		Sets a utility matrix based on empathy, apathy, and yearsInPrison matrix
	Input:
		Player (prisoner): the player
'''
def deriveUtility(Player):
	global yearsInPrison
	Apathy = Player.getApathy()
	Empathy = Player.getEmpathy()
	Years = yearsInPrison
	Utility = []
	Utility.append([])
	Utility.append([])
	Utility[0].append(Years[0][0] * -1 * Apathy	+	Years[0][0] * -1 * Empathy)
	Utility[0].append(Years[0][1] * -1 * Apathy	+	Years[1][0] * -1 * Empathy)
	Utility[1].append(Years[1][0] * -1 * Apathy	+	Years[0][1] * -1 * Empathy)
	Utility[1].append(Years[1][1] * -1 * Apathy	+	Years[1][1] * -1 * Empathy)
	Player.setUtility(Utility)

'''
	Function: getResults
	Description:
		Get the result of a prisoner's dilemma between 2 players
	Input:
		Player1 (prisoner): player 1
		Player2 (prisoner): player 2
	Output:
		return (Dictionary): prison sentences for player 1 and player 2
'''
def getResults(Player1,Player2):
	results = {}
	deriveUtility(Player1)
	deriveUtility(Player2)
	Choice1 = preferRational(Player1)
	Choice2 = preferRational(Player2)
	global yearsInPrison
	results['Player1'] = yearsInPrison[Choice1][Choice2]
	results['Player2'] = yearsInPrison[Choice2][Choice1]
	return results

'''
	Function: getApathyForAction
	Description:
		find all valid apathy values that would cause the specified action
	Input:
		Player (prisoner): the player
		Action (String): the action that would result
	Output:
		return (Number): a valid apathy value
'''
def getApathyForAction(Player,Action):
	global precision
	results = []
	initialApathy = Player.getApathy()
	for i in range(0,precision+1):
		Player.setApathy(i)
		deriveUtility(Player)
		if preferRationalPD(Player) == Action:
			results.append(i)
	Player.setApathy(initialApathy)
	return results

'''
	Function: getEmpathyForAction
	Description:
		find all valid empathy values that would cause the specified action
	Input:
		Player (prisoner): the player
		Action (String): the action that would result
	Output:
		return (Number): a valid empathy value
'''
def getEmpathyForAction(Player,Action):
	global precision
	results = []
	initialEmpathy = Player.getEmpathy()
	for i in range(0,precision+1):
		Player.setEmpathy(i)
		deriveUtility(Player)
		if preferRationalPD(Player) == Action:
			results.append(i)
	Player.setEmpathy(initialEmpathy)
	return results

'''
	Function: maxApathy
	Description:
		select the maximum apathy value for a given action
	Input:
		Player (prisoner): the player
		Action (String): the action that would result
	Output:
		return (Integer): the maximum apathy value
'''
def maxApathy(Player,Action):
	return max(getApathyForAction(Player,Action))
	
'''
	Function: minApathy
	Description:
		select the minimum apathy value for a given action
	Input:
		Player (prisoner): the player
		Action (String): the action that would result
	Output:
		return (Integer): the minimum apathy value
'''
def minApathy(Player,Action):
	return min(getApathyForAction(Player,Action))
	
'''
	Function: maxEmpathy
	Description:
		select the maximum empathy value for a given action
	Input:
		Player (prisoner): identifier for the player
		Action (String): the action that would result
	Output:
		return (Integer): the maximum empathy value
'''
def maxEmpathy(Player,Action):
	return max(getEmpathyForAction(Player,Action))
	
'''
	Function: minEmpathy
	Description:
		select the minimum empathy value for a given action
	Input:
		Player (prisoner): identifier for the player
		Action (String): the action that would result
	Output:
		return (Integer): the minimum empathy value
'''
def minEmpathy(Player,Action):
	return min(getEmpathyForAction(Player,Action))

File: prolog/prisoners.pl
%	prisoners

:- [matrix].

/*
	Rule: setPrecision
	Description:
		set the precision value for the game
	Input:
		Precision (Integer): new value
*/
setPrecision(Precision) :-
	nb_setval(precision,Precision).

/*
	Rule: getPrecision
	Description:
		get the precision value for the game
	Output:
		Precision (Integer): current value
*/
getPrecision(Precision) :-
	nb_getval(precision,Precision).

/*
	Rule: setYearsInPrison
	Description:
		set the yearsInPrison value for the game
	Input:
		YearsInPrison (List): new matrix
*/
setYearsInPrison(YearsInPrison) :-
	nb_setval(yearsInPrison,YearsInPrison).

/*
	Rule: getYearsInPrison
	Description:
		get the yearsInPrison value for the game
	Output:
		YearsInPrison (List): current matrix
*/
getYearsInPrison(YearsInPrison) :-
	nb_getval(yearsInPrison,YearsInPrison).
	
/*
	Rule: setEmpathy
	Description:
		set the empathy value for a player; also implicitly sets the apathy value to its inverse
	Input:
		Player (String): identifier for the player
		Empathy: new value
*/
setEmpathy(Player,Empathy) :-
	getPrecision(Precision),
	setProp(Player,'empathy',Empathy),
	Apathy is Precision - Empathy,
	setProp(Player,'apathy',Apathy).
	
/*
	Rule: getEmpathy
	Description:
		get the empathy value for a player
	Input:
		Player (String): identifier for the player
	Output:
		Empathy: current value
*/
getEmpathy(Player,Empathy) :-
	getProp(Player,'empathy',Empathy).
	
/*
	Rule: setApathy
	Description:
		set the apathy value for a player; also implicitly sets the empathy value to its inverse
	Input:
		Player (String): identifier for the player
		Apathy: new value
*/
setApathy(Player,Apathy) :-
	getPrecision(Precision),
	setProp(Player,'apathy',Apathy),
	Empathy is Precision - Apathy,
	setProp(Player,'empathy',Empathy).
	
/*
	Rule: getApathy
	Description:
		get the apathy value for a player
	Input:
		Player (String): identifier for the player
	Output:
		Apathy: current value
*/
getApathy(Player,Apathy) :-
	getProp(Player,'apathy',Apathy).

/*
	Rule: preferRationalPD
	Description:
		Maps referRational result 0 to 'confess', and result 1 to 'refuse'
	Input:
		Player (String): identifier for the player
	Output:
		Action (String): 'confess' or 'refuse'
*/
preferRationalPD(Player,Action) :-
	deriveUtility(Player),
	preferRational(Player,0),
	Action = 'confess',!;
	preferRational(Player,1),
	Action = 'refuse',!.

/*
	Rule: deriveUtility
	Description:
		Sets a utility matrix based on empathy, apathy, and yearsInPrison matrix
	Input:
		Player (String): identifier for the player
*/
deriveUtility(Player) :-
	getEmpathy(Player,Empathy),
	getApathy(Player,Apathy),
	getYearsInPrison(Years),
	access(Years,0,A),
	access(A,0,A1),
	access(A,1,A2),
	access(Years,1,B),
	access(B,0,B1),
	access(B,1,B2),
	W is A1 * -1 * Apathy + A1 * -1 * Empathy,
	X is A2 * -1 * Apathy + B1 * -1 * Empathy,
	Y is B1 * -1 * Apathy + A2 * -1 * Empathy,
	Z is B2 * -1 * Apathy + B2 * -1 * Empathy,
	Utility = 
	[
		[W,X],
		[Y,Z]
	],
	setUtility(Player,Utility).

/*
	Rule: getResults
	Description:
		Get the result of a prisoner's dilemma between 2 players
	Input:
		Player1 (String): identifier for player 1
		Player2 (String): identifier for player 2
	Output:
		Player1Sentence (Integer): prison sentence for player 1
		Player2Sentence (Integer): prison sentence for player 2
*/
getResults(Player1,Player2,Player1Sentence,Player2Sentence) :-
	deriveUtility(Player1),
	deriveUtility(Player2),
	preferRational(Player1,Choice1),
	preferRational(Player2,Choice2),
	getYearsInPrison(YearsInPrison),
	access(YearsInPrison,Choice1,Player1Row),
	access(Player1Row,Choice2,Player1Sentence),
	access(YearsInPrison,Choice2,Player2Row),
	access(Player2Row,Choice1,Player2Sentence).
	
/*
	Rule: getApathyForAction
	Description:
		find all valid apathy values that would cause the specified action
	Input:
		Player (String): identifier for the player
		Action (String): the action that would result
	Output:
		Apathy (Integer): a valid apathy value
*/
getApathyForAction(Player,Action,Apathy) :-
	getApathy(Player,InitialApathy),
	findall(A,getApathyForActionHelp(Player,Action,A),Results),
	inList(Apathy,Results),
	setApathy(Player,InitialApathy).
	
getApathyForActionHelp(Player,Action,Apathy) :-
	getPrecision(Precision),
	restrict(0,Precision,Apathy),
	setApathy(Player,Apathy),
	deriveUtility(Player),
	preferRationalPD(Player,Action).

/*
	Rule: getEmpathyForAction
	Description:
		find all valid empathy values that would cause the specified action
	Input:
		Player (String): identifier for the player
		Action (String): the action that would result
	Output:
		Empathy (Integer): a valid empathy value
*/
getEmpathyForAction(Player,Action,Empathy) :-
	getEmpathy(Player,InitialEmpathy),
	findall(E,getEmpathyForActionHelp(Player,Action,E),Results),
	inList(Empathy,Results),
	setEmpathy(Player,InitialEmpathy).

getEmpathyForActionHelp(Player,Action,Empathy) :-
	getPrecision(Precision),
	restrict(0,Precision,Empathy),
	setEmpathy(Player,Empathy),
	deriveUtility(Player),
	preferRationalPD(Player,Action).

/*
	Rule: maxApathy
	Description:
		select the maximum apathy value for a given action
	Input:
		Player (String): identifier for the player
		Action (String): the action that would result
	Output:
		Apathy (Integer): the maximum apathy value
*/
maxApathy(Player,Action,Apathy) :-
	findall(Apathy,getApathyForAction(Player,Action,Apathy),Results), %clarify
	max(Results,Apathy).

/*
	Rule: minApathy
	Description:
		select the minimum apathy value for a given action
	Input:
		Player (String): identifier for the player
		Action (String): the action that would result
	Output:
		Apathy (Integer): the minimum apathy value
*/
minApathy(Player,Action,Apathy) :-
	nb_getval(precision,Precision),
	maxEmpathy(Player,Action,Empathy),
	Apathy is Precision - Empathy.

/*
	Rule: maxEmpathy
	Description:
		select the maximum empathy value for a given action
	Input:
		Player (String): identifier for the player
		Action (String): the action that would result
	Output:
		Empathy (Integer): the maximum empathy value
*/
maxEmpathy(Player,Action,Empathy) :-
	findall(Empathy,getEmpathyForAction(Player,Action,Empathy),Results),
	max(Results,Empathy).

/*
	Rule: minEmpathy
	Description:
		select the minimum empathy value for a given action
	Input:
		Player (String): identifier for the player
		Action (String): the action that would result
	Output:
		Empathy (Integer): the minimum empathy value
*/
minEmpathy(Player,Action,Empathy) :-
	nb_getval(precision,Precision),
	maxApathy(Player,Action,Apathy),
	Empathy is Precision - Apathy.
